swarm_dial.go

```diff
@@ -281,39 +286,311 @@ func (s *Swarm) dialPeer(ctx context.Context, p peer.ID) (*Conn, error) {
+
+	var triggerDial <-chan time.Time
+	var nextDial []ma.Multiaddr
+	active := 0
+	done := false
+
+	resch := make(chan dialComplete)
+
+loop:
+	for {
+		select {
+		case req, ok := <-reqch:
+			if !ok {
+				// request channel has been closed, wait for pending dials to complete
+				if active > 0 {
+					done = true
+					reqch = nil
+					triggerDial = nil
+					continue loop
+				}
+
+				// no active dials, we are done
+				return
+			}
+
+			c := s.bestAcceptableConnToPeer(req.Ctx, p)
+			if c != nil {
+				req.Resch <- DialResponse{Conn: c}
+				continue loop
+			}
+
+			addrs, err := s.addrsForDial(req.Ctx, p)
+			if err != nil {
+				req.Resch <- DialResponse{Err: err}
+				continue loop
+			}
+
+			// at this point, len(addrs) > 0 or else it would be error from addrsForDial
+			// ranke them to process in order
+			addrs = s.rankAddrs(addrs)
+
+			// create the pending request object
+			pr := &pendRequest{
+				req:   req,
+				err:   &DialError{Peer: p},
+				addrs: make(map[ma.Multiaddr]struct{}),
+			}
+			for _, a := range addrs {
+				pr.addrs[a] = struct{}{}
+			}
+
+			// check if any of the addrs has been successfully dialed and accumulate
+			// errors from complete dials while collecting new addrs to dial/join
+			var todial []ma.Multiaddr
+			var tojoin []*addrDial
+
+			for _, a := range addrs {
+				ad, ok := pending[a]
+				if !ok {
+					todial = append(todial, a)
+					continue
+				}
+
+				if ad.conn != nil {
+					// dial to this addr was successful, complete the request
+					req.Resch <- DialResponse{Conn: ad.conn}
+					continue loop
+				}
+
+				if ad.err != nil {
+					// dial to this addr errored, accumulate the error
+					pr.err.recordErr(a, ad.err)
+					delete(pr.addrs, a)
+				}
+
+				// dial is still pending, add to the join list
+				tojoin = append(tojoin, ad)
+			}
+
+			if len(todial) == 0 && len(tojoin) == 0 {
+				// all request applicable addrs have been dialed, we must have errored
+				req.Resch <- DialResponse{Err: pr.err}
+				continue loop
+			}
+
+			// the request has some pending or new dials, track it and schedule new dials
+			reqno++
+			requests[reqno] = pr
+
+			for _, ad := range tojoin {
+				ad.requests = append(ad.requests, reqno)
+			}
+
+			if len(todial) > 0 {
+				for _, a := range todial {
+					pending[a] = &addrDial{ctx: req.Ctx, requests: []int{reqno}}
+				}
+
+				nextDial = append(nextDial, todial...)
+				nextDial = s.rankAddrs(nextDial)
+
+				if triggerDial == nil {
+					trigger := make(chan time.Time)
+					close(trigger)
+					triggerDial = trigger
+				}
+			}
+
+		case <-triggerDial:
+			if len(nextDial) == 0 {
+				triggerDial = nil
+				continue loop
+			}
+
+			next := nextDial[0]
+			nextDial = nextDial[1:]
+
+			// spawn the next dial
+			ad := pending[next]
+			go s.dialNextAddr(ad.ctx, p, next, resch)
+			active++
+
+			// select an appropriate delay for the next dial trigger
+			delay := s.delayForNextDial(next)
```

Comment:
Ok, fair enough. I agree with the batching idea.

I'll remove the delay jumping for now and see if I can get the batching working, otherwise I'll just punt.

Reply:
I implemented preliminary batching logic, which can be extended in the future for more finegrained control.
Let me know what you think.

The delay is still there, but it covers batch transitions.

Reply:
I also made it trigger immediately if there are no active dials.

Reply:
I also significantly reduced the delays so that there is no perceptible perf regression.

Reply:
Upon further thought, I decided to remove the delays altogether -- they don't really add anything.

Reply:
Moved the preliminary batching implementation to #252.